design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,291 @@
+   provisioning.
+
+In this scenario, long-lived "pet" clusters that may be more vital to the
+organization are completely managed and published by the infrastructure owner.
+However, for development work, application owners are provided the ability to
+create and delete small clusters as needed using configuration supplied by the
+infrastructure owner (i.e. "clusters as cattle").
+
+#### Scenario 3: Infrastructure owner enables self-service, except for one on-prem cluster with sensitive data
+
+1. Infrastructure owner creates a variety of Kubernetes cluster classes with
+   configuration for different environments and different cloud providers.
+   Application owners are given permissions to create clusters using these
+   classes as they see fit.
+1. There is one particular set of workloads that is required to be run in an
+   existing on-premises cluster (for some important reason). The infrastructure
+   owner creates a Kubernetes `Secret` with connection information to that
+   cluster in an infrastructure namespace (secrets are namespace-scoped). They
+   then create a cluster-scoped Kubernetes `Provider` that references the
+   `Secret`.
+1. Application owner says that they need to run a sensitive workload on that
+   cluster. Infrastructure owner creates a new namespace called
+   `super-secret-workloads` and creates a `KubernetesTarget` in that namespace
+   that references the previously created Kubernetes `Provider`.
+1. Application owner provisions a `KubernetesApplication` with labels that
+   select that `KubernetesTarget` for provisioning.
+
+This example shows the power of this model, but also one of the notable
+shortcomings: if a user can publish clusters to a namespace with a
+`KubernetesTarget`, then they can consume any cluster. For the application
+owners that are doing self-service provisioning in this scenario, they will
+likely need to be able to create `KubernetesTarget` resources for their
+dynamically provisioned Kubernetes cluster managed resources. However, if they
+can do so, they could just as easily create a `KubernetesTarget` for the
+Kubernetes `Provider` that is only to be used for sensitive workloads. While it
+would be nice to restrict the clusters that could be published by users with the
+ability to create `KubernetesTarget` resources, there is not an immediately
+clear and elegant way to do so. It could be argued that only trusted users
+should have this permission level anyway, so it may be less of an issue than
+expressed here. Nevertheless, a potential solution is outlined
+[below](#future-considerations).
+
+## Technical Implementation
+
+The implementation of this proposal is relatively lightweight and requires
+minimal churn on any existing exposed APIs. The first step would be implementing
+the `KubernetesTarget` type in [core Crossplane]. In addition to the creation of
+the custom resource, a new controller would also be required to watch
```

Comment:
I was thinking that it would be a single controller and because the various Kubernetes cluster managed resources support the `Managed` interface which includes [`ConnectionSecretWriterTo`](https://github.com/crossplaneio/crossplane-runtime/blob/bd43e78aa7637acd75c371104bac2cc4778f6e30/pkg/resource/interfaces.go#L74) then the controller could work similar to how the managed resource reconciler interacts with managed resources.

Reply:
It's not clear to me how that would work.

If there was a single controller it would not be able to know that the resource kind it was dealing with was (for example) `*v1alpha3.GKECluster` and thus satisfied the `ConnectionSecretWriterTo` interface, unless the controller imported every possible type of Kubernetes cluster struct from their respective stacks. I believe the controller would instead have to treat the type as `*unstructured.Unstructured`, which does not satisfy the `ConnectionSecretWriterTo` interface.

Reply:
Ah yes this is a good point. It seems like the most straightforward solution would be to implement a lightweight shared reconciler in crossplane-runtime that could be implemented with k8s cluster managed resources in each stack.

Reply:
I think it'd work since `kind` and `apiVersion` is known in `clusterRef`, the controller that watches `WorkloadTarget` could create a `runtime.Object` via https://github.com/crossplaneio/crossplane-runtime/blob/ca4b6b46c9a43f0bf46990ea05f1b23480f8a94f/pkg/resource/resource.go#L75 with that GVK information and then cast that object into `Managed`, extract the secret ref.

Reply:
@muvaf I think you are correct. I was thinking that this should be possible, but was too quick to abandon my original design when confronted with the possibility that it may not be 😅 thanks for pointing this out!

Reply:
Ah, sorry to disappoint 😞  but I realized @negz is right. `MustCreateObject`'s `runtime.ObjectCreater` arguments needs to be aware of the specific struct, which is done usually via stack's `AddToScheme` calls.

I think a controller in `crossplane` repo that uses `*unstructured.Unstructured` makes sense as @negz suggested. We know that `writeConnectionSecretToRef` exists for all managed k8s cluster resources though it exists in all resources that implements `Managed`. So, maybe after getting the secret reference we can validate it's indeed a k8s cluster by looking up a specific key, probably `kubeconfig` and only then propagate the secret.

Reply:
Ah yes, the reason we are able to use `MustCreateObject()` in stacks is due to the fact that we add all the Crossplane types to the scheme. I would push back on the `*unstructured.Unstructured` controller in favor of a lightweight controller for each managed k8s type in stacks. My main reasoning for this is that it feels more like a stack creator intentionally specifying that `KubernetesApplication` resources can be scheduled to a certain resource, which feels in-line with saying that a given resource can be dynamically provisioned by a given claim type.